Fix unit tests for gfx900

whchung opened the pull request:
Fix unit tests on test machines equipped with `gfx900`.

whchung committed:
[test] disable running AMDGPU-specific tests, just compile them

tests change in this commit have been configured to be built with AMDGPU target
gfx803. On test machines with other targets they have no chance to run
successfully. Disable tring to run them but only compile them and check
result IR in this commit.

whchung committed:
[test] add gfx900 into AMDGPU target list

Make these tests pass on test machine equipped with gfx900

scchan commented on tests/Unit/AMDGPU/register-control.cpp:
```diff
@@ -1,6 +1,5 @@
 // RUN: %hc %s -o %t.out -Xlinker -dump-llvm -Xlinker -dump-dir=%T --amdgpu-target=gfx803
 // RUN: %llvm-dis %T/dump-gfx803.opt.bc -f -o - | %FileCheck %s
-// RUN: %t.out
```
should we remove `--amdgpu-target=gfx803` and rely on the enumerator to decide on the arch?

whchung replied:
@scchan I don't like to make `hcc` tests depend on `rocm_agent_enumerator`, as it's another tool not owned/built alongside with `hcc`.  Plus it appears the source code of `rocm_agent_enumerator` isn't posted online so it makes porting it to another architecture hard.

scchan replied:
maybe we should define a new substitution pattern in lit.cfg such that it would replace it with a string including all the GPU architectures?  When we have to add a new architecture in the future, we'll just have to put that in one file.

https://github.com/RadeonOpenCompute/hcc/blob/clang_tot_upgrade/tests/lit.cfg#L142

whchung replied:
@scchan I've thought that, but problem with `rocm_agent_enumerator` is very peculiar:

- unlike most other tools specified in `lit.cfg`, it is NOT owned by `hcc`, so we don't usually know its existence or whereabout.
- unlike `c++filt` which is normally placed under `$PATH`, `rocm_agent_enumerator` is usually installed at `/opt/rocm/bin`, which may usually not be in `$PATH`
- and on a new architecture, `rocm_agent_enumerator` may not even install on `/opt/rocm/bin` and could be altered by `ROCM_ROOT` env var

due to these peculiarities of `rocm_agent_enumerator`, I don't feel like adding an extra dependency to it in `hcc` lit tests.

scchan replied:
my previous comment was about creating something like:

`%target_all_gpus == "--amdgpu-target=gfx701 --amdgpu-target=gfx801 --amdgpu-target=gfx802   --amdgpu-target=gfx900" `

in lit.cfg so that someone could use %target_all_gpus in the unit tests.
It's not about using the enumerator.

whchung replied:
makes sense. let me revise this PR.

scchan commented on tests/Unit/HC/multi_acc2.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %hc --amdgpu-target=gfx701 --amdgpu-target=gfx801 --amdgpu-target=gfx802 --amdgpu-target=gfx803 %s -o %t.out && %t.out
```
maybe we should remove the static list of gpu and just use the enumerator?

whchung replied:
As mentioned in my previous comment I'd like to keep `hcc` be less dependent to `rocm_agent_enumerator`.

whchung committed:
[test] refactor AMDGPU target list into lit.cfg

bensander commented:
perhaps could support a file with the same format as the agent-enumerator, so it could be generated externally by "rocm_agent_enumerator > targets" or manually create target for the alternate archs ?

whchung committed:
[test] fix 2 tests for gfx900

On gfx900 it's not possible to access pointers in host memory range within
kernels. Fix them with HC AM API in this commit.

whchung committed:
[test] fix test case for gfx900

"i" in this test is a host variable, but is captured by reference. it's not
possible on gfx900. fix in this commit.
